static/js/filter.js

```diff
@@ -382,13 +382,48 @@ Filter.prototype = {
     },
 
     can_mark_messages_read: function () {
-        return _.every(this._operators, function (elem) {
-            return (_.contains(['stream', 'topic', 'pm-with'], elem.operator)
-                || elem.operator === 'is' && elem.operand === 'private'
-                || elem.operator === 'in' && elem.operand === 'all'
-                || elem.operator === 'in' && elem.operand === 'home')
-                && !elem.negated;
-        });
+        const term_types = this.sorted_term_types();
+
+        if (_.isEqual(term_types, ['stream', 'topic'])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, ['pm-with'])) {
+            return true;
+        }
+
+        // Folks like me really hate it when Zulip
+        // marks messages as read in interleaved views,
+        // so we will eventually have a setting that
+        // early-exits before the subsequent checks.
+
+        if (_.isEqual(term_types, ['stream'])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, ['is-private'])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, [])) {
+            // All view
+            return true;
+        }
+
+        if (this._operators.length === 1) {
+            // I am not sure why we make special cases for in:all
+            // and in:home, since these are like searches, but we
+            // can fix this in a follow-up.
```

Comment:
I think `in:home` should be removed; I don't think that's a valid narrow.  `in:all` is like the home view with muted streams/topics included, and so the same logic should apply as for the home view.

Reply:
One curiosity here is that we don't handle 'in:all" in our term_type/sorted_term_types scheme.  That was kind of beyond the scope of this commit, but we should probably clean that up.  It should be like a two-line change at its core, but it might have some implications in places that use it.  (I'm gonna assume the history behind this was just oversight on my part when I added sorted_term_types--I don't use in:all personally, so it wasn't on my radar.  Also, I think the changes were kinda experimental at the time.)